delivery/client.py

```diff
@@ -109,9 +132,142 @@ class DeliveryServiceClient:
+
+        :param DeliveryServiceRoutes routes
+            object which contains information of the base url of the desired instance of the
+            delivery-service as well as the available routes
+        :param Iterable[GithubConfig] github_configs (optional):
+            list of the available GitHub configurations which are used to authenticate against the
+            delivery-service
+        :param ConfigFactory cfg_factory (optional):
+            the config factory is used to retrieve available GitHub configurations in case they are
+            not provided anyways (i.e. can be safely omitted in case `github_configs` is specified)
+        '''
         self._routes = routes
-        self.session = requests.sessions.Session()
+        self.github_configs = github_configs
+        self.cfg_factory = cfg_factory
+
+        self._bearer_token = None
+        self._json_web_keys = None
+        self._session = requests.sessions.Session()
+
+    def _openid_configuration(self):
+        res = self._session.get(
+            url=self._routes.openid_configuration(),
+            timeout=(4, 31),
+        )
+
+        res.raise_for_status()
+
+        return res.json()
+
+    def _openid_jwks(self):
+        openid_configuration = self._openid_configuration()
+
+        res = self._session.get(
+            url=openid_configuration.get('jwks_uri'),
+            timeout=(4, 31),
+        )
+
+        res.raise_for_status()
+
+        return res.json()
+
+    def _authenticate(self):
+        if self._bearer_token and not delivery.jwt.is_jwt_token_expired(
+            token=self._bearer_token,
+            json_web_keys=self._json_web_keys,
+        ):
+            return
+
+        res = self._session.get(
+            url=self._routes.auth_configs(),
+            timeout=(4, 31),
+        )
+
+        res.raise_for_status()
+
+        auth_configs = res.json()
+
+        api_urls = [
+            auth_config.get('api_url')
+            for auth_config in auth_configs
+        ]
+
+        github_config = model.github.find_github_config_for_api(
+            api_url=api_urls,
+            github_configs=self.github_configs,
+            cfg_factory=self.cfg_factory,
+        )
+
+        if not github_config:
+            logger.info('no valid credentials found - attempting anonymous-auth')
+            return
+
+        params = {
+            'access_token': github_config.credentials().auth_token(),
+            'api_url': github_config.api_url(),
+        }
+
+        res = self._session.get(
+            url=self._routes.auth(),
+            params=params,
+            timeout=(4, 31),
+        )
+
+        if not res.ok:
+            logger.warning(
+                'authentication against delivery-service failed: '
+                f'{res.status_code=} {res.reason=} {res.content=}'
+            )
+
+        res.raise_for_status()
+
+        self._bearer_token = res.cookies.get(delivery.jwt.JWT_KEY)
```

Comment:
should we check such a cookie was actually returned (and emit an error if this was not the case)?